src/claimtriedb.h

```diff
@@ -0,0 +1,140 @@
+#ifndef BITCOIN_CLAIMTRIE_DB_H
+#define BITCOIN_CLAIMTRIE_DB_H
+
+#include "dbwrapper.h"
+
+#include <utility>
+
+#include <map>
+
+/**
+ * additional macros to complete hash algorithm
+ */
+#define  len(s)          (sizeof(s)/sizeof(*s)-1)
+#define   H1(s, i, x)    (x*65599u+(uint8_t)s[(i)<len(s)?len(s)-1-(i):len(s)])
+#define   H4(s, i, x)  H1(s, i,  H1(s, i+ 1,  H1(s, i+  2,  H1(s, i+  3, x))))
+#define  H16(s, i, x)  H4(s, i,  H4(s, i+ 4,  H4(s, i+  8,  H4(s, i+ 12, x))))
+#define  H64(s, i, x) H16(s, i, H16(s, i+16, H16(s, i+ 32, H16(s, i+ 48, x))))
+#define H256(s, i, x) H64(s, i, H64(s, i+64, H64(s, i+128, H64(s, i+192, x))))
+
+/**
+ * Compile time hash calculation algorithm
+ * @param[in] s    compile time string to compute its hash
+ */
+#define HASH(s)          ((uint32_t)((H256(s, 0, 0) ^ (H256(s, 0, 0) >> 16))))
+
+/**
+ * function to provide unique hash of 2 types
+ */
+template <class T1, class T2>
+size_t hashType();
+
+/**
+ * explicit template specialization,
+ * it should be provided where it's need a type hash
+ */
+#define HASH_VALUE(T1, T2) \
+template<> inline          \
+size_t hashType<T1, T2>()  \
+{                          \
+    return HASH(#T1 #T2);  \
+}
```

Comment:
If we really want to do this, there are plenty of template specialization varieties that can be made that resort to std::hash which eliminates the need for many of the custom hash variants.

Reply:
Yes, it's helper macro. There is no need to write class names by hand.